ioSPI/fourier.py

```diff
@@ -0,0 +1,237 @@
+import numpy as np
+import pyfftw.interfaces.numpy_fft
+from numba import jit
+
+
+@jit
+def make_neg_pos_2d(arr2d_shape):
+    """Apply checkerboard pattern for 2D FFT.
+
+    Each pixel switches from positive to negative in checker board pattern.
+    Equivalent to fft shifting
+    https://dsp.stackexchange.com/questions/9039/centering-zero-frequency-for-discrete-fourier-transform
+
+    Parameters
+    ----------
+    arr2d_shape : tuple (n_particle,N_fourier_pixels,N_fourier_pixels)
+        Input shape tuple.
+
+    Returns
+    -------
+    neg_pos_2d : numpy.ndarray, shape (n_particle,N_fft_pix,N_fft_pix)
+        Checkerboard array.
+    """
+    # assert arr2d.ndim == 3  # extra axis
+    n_particle, R1, R2 = arr2d_shape
+    neg_pos_2d = np.ones(arr2d_shape)
+    for particle in range(n_particle):
```

Comment:
These 3 for loops look like they could be run in parallel. Any chance we could use joblib there?

Reply:
Or maybe this solution could be faster:
```
def checkerboard(shape):
...     return 2 * (np.indices(shape).sum(axis=0) % 2 - 0.5 * np.ones(shape))
```

while also working for any dim of arrays?

Reply:
Its numba jitted, so it's spedup and has parallelized the loops.

But since there is a way to vectorize it, I benchmarked.

```
def checkerboard(shape):
  return -2*(np.indices(shape).sum(axis=0) % 2 - 0.5)

for num_pixels in [8,16,32,36,128,256,512]:
  print(num_pixels)
  %timeit checkerboard((num_pixels,num_pixels))
  %timeit fourier.make_neg_pos_2d((1,num_pixels,num_pixels))

8
The slowest run took 76.79 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 16.9 µs per loop
The slowest run took 33.43 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 5: 1.09 µs per loop
16
The slowest run took 5.39 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 5: 19.4 µs per loop
The slowest run took 14.21 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 5: 1.33 µs per loop
32
The slowest run took 48.22 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 5: 31.4 µs per loop
The slowest run took 8.31 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 2.46 µs per loop
36
The slowest run took 19.27 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 5: 35.3 µs per loop
The slowest run took 6.95 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 3.22 µs per loop
128
The slowest run took 8.80 times longer than the fastest. This could mean that an intermediate result is being cached.
1000 loops, best of 5: 256 µs per loop
10000 loops, best of 5: 26.8 µs per loop
256
1000 loops, best of 5: 1.28 ms per loop
10000 loops, best of 5: 143 µs per loop
512
100 loops, best of 5: 5.11 ms per loop
1000 loops, best of 5: 636 µs per loop
```

Numba is faster.

Reply:
Resolved.